aip/0194.md

```diff
@@ -0,0 +1,237 @@
+- Non-retryable: requests which **should not** be retried. Examples include
+  CRUD operations which don't take a unique ID, most SQL operations, and
+  operations that alter state when retried (for example,
+  `IncrementBy(some-uuid, some-field, 7)`, which takes a unique ID but will
+  always alter state).
+
+Streaming requests, on the other hand, can be further categorized as:
+
+- Server streaming: requests in which the server streams data to the client.
+  These involve no data sent from the client, meaning that it generally is an
+  idempotent `GET`. That makes it equivalent to a retryable unary request as far
+  as retrying error codes go.
+- Client streaming: requests in which the client streams data to the server. The
+  retryability of these requests is highly dependent on the type of data being
+  streamed. This document does not provide guidance for these requests.
+- Bidirectional: requests in which both the client streams to the server, and
+  the server streams to the client. The retryability of these requests is
+  highly dependent on the type of data being streamed. This document does not
+  provide guidance for these requests.
+
+### Retryable codes
+
+These codes **should** be retried for retryable requests:
+
+- `INTERNAL`: This generally means that some internal part of the system has
+failed. These are retryable under the expectation that the system will recover
+from failures.
+- `UNAVAILABLE`: These codes generally result from network hiccups, and are
+generally transient. They are retryable under the expectation that the
+connection will become available (soon).
+
+### Non-retryable codes
+
+These codes **should** never be retried for any request:
+
+- `OK`: The request succeeded.
+- `CANCELLED`: A user can cancel a request, which **must** be honoured.
+- `DEADLINE_EXCEEDED`: A user can set a deadline, which **must** be honoured.
+- `INVALID_ARGUMENT`: Retrying a request with an invalid argument will never
+  succeed.
+- `DATA_LOSS`: This is an unrecoverable error and **must** immediately be
+  surfaced to the user.
+
+### Generally non-retryable codes:
+
+Some codes can be retried if a system is designed without synchronization
+between various components. For example, a system in which a read is designed to
+hang forever until a resource is created. However, these are not the types of
+systems generally encouraged, and a client **should not** retry these types of
+errors.
+
+These codes generally **should not** be retried for any request:
+
+- `RESOURCE_EXHAUSTED`: This may be a signal that the user's quota is exhausted.
+  Retries therefore may not be expected to work for several hours; meanwhile the
+  retries may have billing implications. If `RESOURCE_EXHAUSTED` is used for
+  other reasons than quota and the expected time for the resource to become
+  available is much shorter, it may be retryable.
+- `NOT_FOUND`: The resource may be created in the future.
```

Comment:
It's not clear what these "X may happen in the future" statements are intended to achieve. Are they justifications for not retrying in general? Or suggestions about when they might be retryable in special cases?

Reply:
Generally concur. I think what we are trying to communicate here is that the state is _not_ likely to be different if you try again. If something returns `NOT_FOUND` and you retry it, it probably still will not be found.

Reply:
Done

Reply:
See https://github.com/jadekler/aip/blob/194/aip/0194.md